00-config-refactor.md

```diff
@@ -0,0 +1,102 @@
+# Proposal: Config Refactor
+
+## Background
+
+For historical reasons, our SDK has the following problems:
+
+- SDK's config path is hardcoded to `~/.config/qingstor.yaml`, if not exist, SDK will create automatically.
+- There is not way for user to change SDK's config path during runtime. This enforces all tools to implement their own
+  SDK loading logic.
+- SDK's assume all user have access key pairs which forbids anonymous API call.
+- SDK use virtual host style to check bucket but use path style for API call.
+- SDK can't control IPv4/IPv6 logic
+
+## Proposal
+
+So I propose to do following config refactors:
+
+### Environment First Config
+
+Instead of hardcoded `~/.config/qingstor.yaml`, we will read config path from system environment
+variable `QINGSTOR_CONFIG_PATH`.
+
+If `QINGSTOR_CONFIG_PATH` is not exists, we will fallback to `~/.config/qingstor.yaml`.
+
+And all SDKs should allow reading config from envs:
+
+```shell
+export QINGSTOR_ACCESS_KEY_ID=example_access_key_id
+export QINGSTOR_SECRET_ACCESS_KEY=example_secret_access_key
+export QINGSTOR_XYZ=example_xyz
+```
+
+For every config, we should check via following order:
+
+- The specific env: `QINGSTOR_ACCESS_KEY_ID`
+- Config loading from `QINGSTOR_CONFIG_PATH`
+- If no default value, keep empty, or use the default value.
+
+All qingstor native tools could use the same logic here.
+
+### Merge Host, Port, Protocol into Endpoint
+
+Instead of use split `host`, `port`, `protocol` config values, we will introduce a new config: `endpoint`.
+
+```yaml
+host: qingstor.com
+port: 443
+protocol: https
+```
+
+equals to
+
+```yaml
+endpoint: https://qingstor.com:443
+```
+
+> Especially for `http://xxx.com:80` && `https://xxx.com:443`, the tailing port number could be omitted.
+
+If user has set `endpoint`, `host`, `port` and `protocol` should be ignored with a warning printed for user.
+
+The default value for `endpoint` is `https://qingstor.com`
+
+### Virtual Host and Path Style support
+
+SDK should support both `virtual host style` and `path style`.
+
+We will introduce a new config: `enable_virtual_host_style`. If set to `true`, SDK will use `virtual host style` for all API calls.
```

Comment:
this config item is named `virtual_host_enabled` in java sdk.
Maybe a verb-like name is not good for a boolean variable?

Reply:
It's a naming style from: https://github.com/aws/aws-sdk-go/blob/master/aws/config.go#L127

Compared to `abc_disabled`, `abd_forced`, `abc_enabled`, I found `disable_abc`, `force_abc`, `enable_abc` more readable.

Reply:
Agreed. I will migrate to this style.

Reply:
I think configs like `enable_virtual_host_style: true` in `yaml` file is more readable. But as an env variable, `VIRTUAL_HOST_ENABLED` is better.
However, if we want to follow the rule to keep consistence between `QINGSTOR_XYZ` (env) and `xyz` (in yaml), we have to use `enable_virtual_host_style` instead of `VIRTUAL_HOST_ENABLED`.